cassandra/policies.py

```diff
@@ -316,6 +327,138 @@ def on_add(self, host):
     def on_remove(self, host):
         self.on_down(host)
 
+class RackAwareRoundRobinPolicy(LoadBalancingPolicy):
+    """
+    Similar to :class:`.RackAwareRoundRobinPolicy`, but prefers hosts
+    in the local rack, before hosts in the local datacenter but a
+    different rack, before hosts in all other datercentres
+    """
+
+    local_dc = None
+    local_rack = None
+    used_hosts_per_remote_dc = 0
+
+    def __init__(self, local_dc='', local_rack='', used_hosts_per_remote_dc=0):
+        """
+        The `local_dc` and `local_rack` parameters should be the name of the
+        datacenter and rack (such as is reported by ``nodetool ring``) that
+        should be considered local. If not specified, the driver will choose
+        a both based on the first host among :attr:`.Cluster.contact_points`
+        having a valid DC and rack. If relying on this mechanism, all specified
+        contact points should be nodes in a single, local DC.
+
+        `used_hosts_per_remote_dc` controls how many nodes in
+        each remote datacenter will have connections opened
+        against them. In other words, `used_hosts_per_remote_dc` hosts
+        will be considered :attr:`~.HostDistance.REMOTE` and the
+        rest will be considered :attr:`~.HostDistance.IGNORED`.
+        By default, all remote hosts are ignored.
+        """
+        self.local_rack = local_rack
+        self.local_dc = local_dc
+        self.used_hosts_per_remote_dc = used_hosts_per_remote_dc
+        self._live_hosts = {}
+        self._position = 0
+        self._endpoints = []
+        LoadBalancingPolicy.__init__(self)
+
+    def _rack(self, host):
+        return host.rack or self.local_rack
+
+    def _dc(self, host):
+        return host.datacenter or self.local_dc
+
+    def populate(self, cluster, hosts):
+        for (dc, rack), dc_hosts in groupby(hosts, lambda h: (self._dc(h), self._rack(h))):
+            self._live_hosts[(dc, rack)] = tuple(set(dc_hosts))
+
+        if not self.local_rack and not self.local_dc:
+            self._endpoints = [
+                endpoint
+                for endpoint in cluster.endpoints_resolved]
+
+        self._position = randint(0, len(hosts) - 1) if hosts else 0
```

Comment:
I think a few word regarding this internal implementation, and what this randomness is for, could be helpful.

I couldn't figure it from the code

Reply:
o.k. would still recommend to document it, even if upstream didn't